### PR TITLE
My plan: Add plugin setup

### PR DESCRIPTION
**Pull request by sirreal**
- Send Jetpack purchases to `/plan/my-plan/SITE_SLUG/?do-setup` after checkout.
- When users are on `/plan/my-plan/SITE_SLUG/?do-setup` with a Jetpack, non-Atomic site with a paid purchase, install Akismet and Vaultpress and provision the plan-provided keys.
- Add a plan plugin key selector

This was broken out from #26690 to focus on the plan setup.

This is largely a replacement for the functionality in on the Jetpack-specific checkout thank you page. It was problematic to re-use or translate that logic for a few reasons:
- UI and the installation process were tightly coupled.
- The Plugin Store exists largely outside of React.
- Redux and the component lifecycle drive the installation flow incidentally rather than intentionally.

This PR extracts some [plugin installer logic from the WooCommerce extension](https://github.com/Automattic/wp-calypso/blob/fc5a5e5141d94162e90936af074c85853ff43e0e/client/extensions/woocommerce/app/dashboard/required-plugins-install-view.js
). The logic has been cleaned up to focus solely on Plugin installation. A nice follow-up would be to share this component between these use cases. This installer exists as a state machine that repeatedly checks Plugin installation progress to proceed through the flow. Reporting has been added to the installer to allow parents to listen to installation process.

This PR also adds plan provisioning logic (set up plugin keys) to a component. This process is driven by state updates and the component lifecycle. The component managing provisioning listens to the Plugin installation process, and when it is complete, will proceed to sequentially provision Akismet and VaultPress keys. This component can also notify of its progress. At this time, that is limited to complete and total tasks, but more detailed state updates can be added in follow-up changes as needed by the UI (#26881)

Much of this logic existed in some way due to the Jetpack checkout thank you page. However, that logic is a rather convoluted way of updating state that is only of interest to a small component tree during a limited time. Using the Redux state flow to drive a flow through a small component hierarchy's lifecycle method spreads out a lot of logic that makes sense to be co-located.

Follow conversation here: p7rd6c-1R4-p2

## Follow-up work
- Warn on exit if process is incomplete
- Error handling (some messaging and error conditions exist in from the thank you page flow)
- Better reporting of state to parents
- Remove `?do-setup` from the URL on completion.
- The checklist steps should be connected to state that it indicative of the process at a high level. This will probably come from the installed plugins state tree. I'll need to investigate whether we have a way of checking whether plugins are correctly provisioned with keys from Calypso or not. We may not need to get that specific.
- _Nice to have_: Relocate plugin installer component and share back to WooCommerce extension
- _Nice to have_: Update purchases view to use new plugin key selector

## Testing

_All testing should be available via [Calypso live](https://calypso.live/jetpack/connect?branch=add/plan-setup)_

1. Get debug output:
   `localStorage.debug = 'calypso:jetpack-setup-runner'`
   There isn't much debug output, but you will see as plugins are provisioned (error or success). That will tell you when the process is complete. Remember to reload after setting debug 👍 
1. Connect a new Jetpack site
1. Purchase a paid plan for the site
1. You should be sent to `/plan/my-plan/SITE_SLUG/?do-setup` after checkout
1. There is no UI, but if you apply the diff below you will be able to see the state as the process is followed
1. So far so good?
1. If you haven't applied the diff, you won't see any UI and since nothing is rendered, it's difficult to inspect React state internally. However, you can verify that VaultPress is working on your site via WP Admin (Jetpack -> VaultPress menu) and that Akismet has a key (Jetpack -> Akismet menu).
1. Break stuff!
   - Not an admin user
   - Can't edit file system
   - Can't `manage_options`
   - Plugin already installed
   - Old versions of VaultPress (`<=1.8.3`) - provision correctly? How many folks actually use this?
   - You name it!

<details>
<summary>Use the diff below to dump raw props/state into rendered HTML</summary>

```diff
diff --git a/client/my-sites/plans/current-plan/index.jsx b/client/my-sites/plans/current-plan/index.jsx
index b7fb69f82c..dd8d04f5ee 100644
--- a/client/my-sites/plans/current-plan/index.jsx
+++ b/client/my-sites/plans/current-plan/index.jsx
@@ -125,11 +125,14 @@ class CurrentPlan extends Component {
 				<QuerySitePlans siteId={ selectedSiteId } />
 				{ shouldQuerySiteDomains && <QuerySiteDomains siteId={ selectedSiteId } /> }
 				{ doJetpackPlanSetup && (
-					<JetpackSetupRunner
-						key={ /* Force remount on site change */ selectedSiteId }
-						notifyProgress={ this.updatePlanSetupProgress }
-						siteId={ selectedSiteId }
-					/>
+					<>
+						<JetpackSetupRunner
+							key={ /* Force remount on site change */ selectedSiteId }
+							notifyProgress={ this.updatePlanSetupProgress }
+							siteId={ selectedSiteId }
+						/>
+						<pre>{ JSON.stringify( this.state.planSetup, undefined, 2 ) }</pre>
+					</>
 				) }
 
 				<PlansNavigation path={ path } selectedSite={ selectedSite } />
diff --git a/client/my-sites/plans/current-plan/jetpack-setup-runner/index.js b/client/my-sites/plans/current-plan/jetpack-setup-runner/index.js
index d53bb883f0..6b527c31ca 100644
--- a/client/my-sites/plans/current-plan/jetpack-setup-runner/index.js
+++ b/client/my-sites/plans/current-plan/jetpack-setup-runner/index.js
@@ -185,17 +185,27 @@ class JetpackSetupRunner extends PureComponent {
 	render() {
 		const { siteId } = this.props;
 		return (
-			siteId && (
-				<>
-					<QueryPluginKeys siteId={ siteId } />
-					<DoPluginSetup
-						key={ /* Force remount on site change */ siteId }
-						notifyProgress={ this.handleUpdateProgress }
-						requiredPlugins={ [ 'akismet', 'vaultpress' ] }
-						siteId={ siteId }
-					/>
-				</>
-			)
+			<>
+				{ siteId && (
+					<>
+						<QueryPluginKeys siteId={ siteId } />
+						<DoPluginSetup
+							key={ /* Force remount on site change */ siteId }
+							notifyProgress={ this.handleUpdateProgress }
+							requiredPlugins={ [ 'akismet', 'vaultpress' ] }
+							siteId={ siteId }
+						/>
+					</>
+				) }
+				<h1>props:</h1>
+				<pre>
+					<code>{ JSON.stringify( this.props, undefined, 2 ) }</code>
+				</pre>
+				<h1>state:</h1>
+				<pre>
+					<code>{ JSON.stringify( this.state, undefined, 2 ) }</code>
+				</pre>
+			</>
 		);
 	}
 }
```
</details>

### CODE REVIEW

File: client/my-sites/checkout/checkout/checkout.jsx [hunk 1]
```diff
@@ -357,6 +358,10 @@ export class Checkout extends React.Component {
 			}
 		}
 
+		if ( this.props.isJetpack && isEnabled( 'jetpack/checklist' ) ) {
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/checkout/checkout/checkout.jsx]: 1ff6e022815137c04dc412b16cc02e961c0fb142

File: client/my-sites/checkout/checkout/checkout.jsx [hunk 2]
```diff
@@ -357,6 +358,10 @@ export class Checkout extends React.Component {
...
+			return `/plans/my-plan/${ selectedSiteSlug }?do-setup`;
+		}
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/checkout/checkout/checkout.jsx]: 7ee02832cef57d3c8a7777f1dd108b57a1ddfbbc

File: client/my-sites/plans/current-plan/index.jsx [hunk 3]
```diff
@@ -113,6 +120,12 @@ class CurrentPlan extends Component {
 				<QuerySites siteId={ selectedSiteId } />
 				<QuerySitePlans siteId={ selectedSiteId } />
 				{ shouldQuerySiteDomains && <QuerySiteDomains siteId={ selectedSiteId } /> }
+				{ doPlanSetup && (
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/index.jsx]: 0334a848c7b2d37899ede6122f523dedaf233b0f

File: client/my-sites/plans/current-plan/jetpack-setup-runner/index.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+const KEY_PROVISION_STATE_DONE = 'KPS_DONE';
+const KEY_PROVISION_STATE_FAIL = 'KPS_FAIL';
+const KEY_PROVISION_STATE_IDLE = 'KPS_IDLE';
+const KEY_PROVISION_STATE_IN_PROGRESS = 'KPS_IN_PROGRESS';
+
+class JetpackSetupRunner extends PureComponent {
+	static propTypes = {
+		site: PropTypes.shape( {
+			ID: PropTypes.number.isRequired,
+		} ),
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/index.js]: d531a07dc6ae08bbc47e72a05ec4255d313d2318

File: client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+export const ENGINE_STATE_DONE_FAIL = 'ES_DONE_FAIL';
+export const ENGINE_STATE_DONE_SUCCESS = 'ES_DONE_SUCCESS';
+export const ENGINE_STATE_INITIALIZE = 'ES_INIT';
+export const ENGINE_STATE_INSTALL = 'ES_INSTALL';
+
+class PluginInstaller extends Component {
+	static propTypes = {
+		site: PropTypes.shape( {
+			ID: PropTypes.number.isRequired,
+		} ),
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js]: 55180583874d1ad4599d6d0804a931f19ad66fe1

File: client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+/**
+ * Internal dependencies
+ */
+import { activatePlugin, installPlugin, fetchPlugins } from 'state/plugins/installed/actions';
+import { fetchPluginData } from 'state/plugins/wporg/actions';
+import { getPlugin } from 'state/plugins/wporg/selectors';
+import { getPlugins, getStatusForSite } from 'state/plugins/installed/selectors';
+import { getSelectedSite } from 'state/ui/selectors';
+
+const debug = debugFactory( 'calypso:plugin-setup' ); // eslint-disable-line no-unused-vars
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js]: It was unused. Removed.

File: client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+	};
+
+	render() {
+		return null;
+	}
+}
+
+function mapStateToProps( state ) {
+	const site = getSelectedSite( state );
+	const siteId = site.ID;
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/do-plugin-setup.js]: 55180583874d1ad4599d6d0804a931f19ad66fe1

File: client/my-sites/plans/current-plan/jetpack-setup-runner/index.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+				KEY_PROVISION_STATE_IDLE === this.state.vaultpressKeyProvisioning &&
+				( KEY_PROVISION_STATE_DONE === this.state.akismetKeyProvisioning ||
+					KEY_PROVISION_STATE_FAIL === this.state.akismetKeyProvisioning )
+			) {
+				this.provisionVaultpressKey();
+			}
+		}
+	}
+
+	provisionAkismetKey() {
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/index.js]: 3a4f991a3e83773062a0b6f60034c8bc37eff446

File: client/my-sites/plans/current-plan/jetpack-setup-runner/index.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+	componentDidUpdate( prevProps, prevState ) {
+		this.maybeProvisionNextKey();
+		this.maybeReportProgress( prevState );
+	}
+
+	/**
+	 * React to state updates that should initiate and proceed through key provisioning
+	 */
+	maybeProvisionNextKey() {
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/jetpack-setup-runner/index.js]: Good question and needs some further investigation.

- The current (Jetpack thank you page) implementation would do the same thing. This is protected by the fact that it's difficult to get back to the thank you page after provisioning has run. Some existing redux state state does represent whether provisioning is done or not, but I believe the thank you implementation still just goes ahead and tries to do everything.
- This implementation would rely on pretty much the same thing, relying on the existence of the `?do-setup` GET param to trigger running the set up. Most routing would not link to this, and after setup we'd redirect without the `?do-setup` param. We might need to include some stateful setup params, but that can be iterated on.
- To the best of my knowledge, no API exists explicitly to answer the question of whether plugin keys are provisioned. We'd need to query the site options. I was starting work in #26810, but querying site options appears to be extremely slow. We should avoid putting keys (and thus these options) into persisted state, but there is work required here, some of which is noted in the follow-up section of the description:
  > I'll need to investigate whether we have a way of checking whether plugins are correctly provisioned with keys from Calypso or not. 

If we just re-provision with working keys, that might be the best outcome, we could use some VaultPress expertise there.

There are a few scenarios:

- Site has no working key. Provisioning is good 👍 
- Site already has the to-provision key provisioned. Redundant work, doesn't seem to be a problem 🤷‍♂️ 
- Site already has a different provisioned key, and we provision a new valid key. No idea what the impact would be.
  @singerb - If a site has a working VaultPress key and it's swapped for another working key, does that have any impact for VaultPress? Would that potentially change the owner of the backups or have other negative impact?

**Comment by sirreal** [comment]
Conversation moved here: p7rd6c-1R4-p2

File: client/my-sites/checkout/checkout/checkout.jsx [hunk 10]
```diff
@@ -667,6 +674,8 @@ export default connect(
 			isPlansListFetching: isRequestingPlans( state ),
 			isSitePlansListFetching: isRequestingSitePlans( state, selectedSiteId ),
 			planSlug: getUpgradePlanSlugFromPath( state, selectedSiteId, props.product ),
+			isJetpackNotAtomic:
```
[the remark this answers is not in the document]
Comment (ockham) [reply client/my-sites/checkout/checkout/checkout.jsx]: Then again: p4TIVU-7pt-p2 :grin: 

> ## Should Atomic sites receive special treatment?
> No! [...]

File: client/my-sites/plans/current-plan/index.jsx [hunk 11]
```diff
@@ -37,6 +37,8 @@ import SidebarNavigation from 'my-sites/sidebar-navigation';
 import JetpackChecklist from 'my-sites/plans/current-plan/jetpack-checklist';
 import { isEnabled } from 'config';
 import QueryJetpackPlugins from 'components/data/query-jetpack-plugins';
+import JetpackSetupRunner from 'my-sites/plans/current-plan/jetpack-setup-runner/index';
```
[the remark this answers is not in the document]
Comment (ockham) [reply client/my-sites/plans/current-plan/index.jsx]: Should even work without renaming, shouldn't it?

File: client/my-sites/plans/current-plan/index.jsx [hunk 12]
```diff
@@ -113,6 +124,13 @@ class CurrentPlan extends Component {
 				<QuerySites siteId={ selectedSiteId } />
 				<QuerySitePlans siteId={ selectedSiteId } />
 				{ shouldQuerySiteDomains && <QuerySiteDomains siteId={ selectedSiteId } /> }
+				{ doJetpackPlanSetup && (
+					<JetpackSetupRunner
+						key={ /* Force remount on site change */ selectedSiteId }
```
Comment (ockham) [inline client/my-sites/plans/current-plan/index.jsx]: Hmm, shouldn't that be handled by a lifecycle method in `JetpackSetupRunner` that watches out for changes of `siteId`?

File: client/my-sites/checkout/checkout/checkout.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/checkout/checkout/checkout.jsx]: I would rephrase a few things there, but I do try to keep that in mind 🙂 

It really comes down to _all_ site types needing _equal_ treatment.

File: client/my-sites/plans/current-plan/index.jsx (hunk 12 above)
Comment (sirreal) [reply client/my-sites/plans/current-plan/index.jsx]: It's pretty difficult to properly reinitialize components that aren't remounted. This is a way of ensuring that, if the siteId were changed, the component would be thrown out and mounted fresh– hat tip @jsnajdr p4TIVU-910-p2

File: client/my-sites/plans/current-plan/index.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/my-sites/plans/current-plan/index.jsx]: It works fine, including it was just a mistake on my part. Strong convention is not to include the index for better or worse.

**Comment by sirreal** [comment]
We'll try a different approach that doesn't sidestep so many perceived issues.

Thanks everyone!
